File: api/models/ticket.py
"""
Ticket model – represents a booked flight ticket.

Schema reference: docs/ARCHITECT.md § Model: Ticket
Dictionary:       docs/DICTIONARY.md  (pnr, net_price, ev_price, ast_price, thf_price, web_price, insurance_price, selling_price, discount, true_income, itinerary)
Agent output:     docs/AGENT_PARSER.md (fields produced by the AI extraction step)

True income per ticket = selling_price + discount - (ev_price + ast_price + thf_price + web_price + insurance_price)
"""

import uuid
from datetime import datetime, timezone
from typing import TYPE_CHECKING, List, Optional

from sqlalchemy import Column
from sqlmodel import Field, Relationship, SQLModel

try:
    from sqlalchemy.types import JSON
except ImportError:  # pragma: no cover
    from sqlalchemy import JSON  # type: ignore

from .enums import Airline, TicketStatus

if TYPE_CHECKING:
    from .customer import Customer
    from .invoice_item import InvoiceItem
    from .quote_item import QuoteItem
    from .transaction import Transaction


# ---------------------------------------------------------------------------
# Shared / base properties
# ---------------------------------------------------------------------------

class TicketBase(SQLModel):
    """Fields shared between create/read schemas and the DB table."""

    # Mã đặt chỗ – optional 6-character booking reference shared by grouped passengers.
    pnr: Optional[str] = Field(
        default=None,
        index=True,
        nullable=True,
        min_length=6,
        max_length=6,
        description=(
            "Optional 6-character PNR (Passenger Name Record) booking reference code. "
            "May repeat across passenger rows in group bookings."
        ),
    )
    airline: Optional[Airline] = Field(
        default=None,
        description="Carrier code: VNA (Vietnam Airlines), VJ (Vietjet), QH (Bamboo), VU (Vietravel).",
    )
    ticket_number: Optional[str] = Field(
        default=None,
        index=True,
        max_length=50,
        description="Airline ticket number. May repeat across outbound/return ticket rows.",
    )
    seat_code: Optional[str] = Field(
        default=None,
        max_length=20,
        description="Optional seat assignment code, e.g. 12A.",
    )
    fare_class: Optional[str] = Field(
        default=None,
        max_length=50,
        description="Optional fare class / fare family label from the source ticket, e.g. B or Flexible.",
    )
    departure_place: Optional[str] = Field(
        default=None,
        max_length=255,
        description="Human-readable departure place, e.g. Da Nang City.",
    )
    arrival_place: Optional[str] = Field(
        default=None,
        max_length=255,
        description="Human-readable arrival place, e.g. Ho Chi Minh City.",
    )
    departure_code: Optional[str] = Field(
        default=None,
        max_length=10,
        description="Compact departure place code, e.g. DAD.",
    )
    arrival_code: Optional[str] = Field(
        default=None,
        max_length=10,
        description="Compact arrival place code, e.g. SGN.",
    )
    # Hành trình – e.g. "HAN-SGN"
    itinerary: Optional[str] = Field(
        default=None,
        max_length=100,
        description='Flight route string (hành trình), e.g. "HAN-SGN" or "SGN-DAD-HAN".',
    )
    flight_date: datetime = Field(
        description="Scheduled departure datetime stored in ISO-8601 / UTC.",
    )
    booked_at: Optional[datetime] = Field(
        default=None,
        index=True,
        description="Real-world datetime when the ticket was booked manually by staff.",
    )
    # Giá gốc – cost paid to the airline / supplier.
    net_price: float = Field(
        ge=0,
        description="Net cost from airline/supplier (giá gốc). Must be ≥ 0.",
    )
    ev_price: float = Field(
        default=0.0,
        ge=0,
        description="Host price from EV (giá EV). Empty values count as 0.",
    )
    ast_price: float = Field(
        default=0.0,
        ge=0,
        description="Host net price from AST (giá AST). Empty values count as 0.",
    )
    thf_price: float = Field(
        default=0.0,
        ge=0,
        description="Host net price from Thanh Hoang / THF (giá Thành Hoàng). Empty values count as 0.",
    )
    web_price: float = Field(
        default=0.0,
        ge=0,
        description="Host net price from WEB (giá WEB). Empty values count as 0.",
    )
    insurance_price: float = Field(
        default=0.0,
        ge=0,
        description="Insurance price (giá bảo hiểm). Empty values count as 0.",
    )
    # Giá bán – price invoiced to the customer.
    selling_price: float = Field(
        ge=0,
        description="Selling price charged to the customer (giá bán). Must be ≥ 0.",
    )
    discount: float = Field(
        default=0.0,
        ge=0,
        description="Airline add-in / discount amount earned by the agency for this ticket in VND.",
    )
    true_income: float = Field(
        default=0.0,
        description="Actual ticket income: selling_price + discount - (ev_price + ast_price + thf_price + web_price + insurance_price).",
    )

    status: TicketStatus = Field(
        default=TicketStatus.DRAFT,
        description="Lifecycle state of the ticket (DRAFT, CONFIRMED, VOID, REFUNDED).",
    )

    # FK – resolved at the DB layer.
    customer_id: uuid.UUID = Field(foreign_key="customer.id", index=True)


# ---------------------------------------------------------------------------
# DB Table
# ---------------------------------------------------------------------------

class Ticket(TicketBase, table=True):
    """Persisted ticket entity stored in the `ticket` table."""

    __tablename__ = "ticket"

    id: Optional[uuid.UUID] = Field(
        default_factory=uuid.uuid4,
        primary_key=True,
        index=True,
        nullable=False,
    )

    # passengers stored as a JSON array of uppercase full-name strings.
    # e.g. ["NGUYEN VAN A", "TRAN THI B"]
    passengers: List[str] = Field(
        default_factory=list,
        sa_column=Column(JSON, nullable=False, default=list),
        description="List of passenger full names (UPPERCASE), stored as JSON.",
    )
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        index=True,
        description="UTC timestamp of when this ticket record was created.",
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        index=True,
        description="UTC timestamp of when this ticket record was last updated.",
    )

    # Relationship back to Customer.
    customer: Optional["Customer"] = Relationship(back_populates="tickets")

    # Relationship to Transactions created from this ticket (e.g. auto-debt).
    transactions: List["Transaction"] = Relationship(back_populates="linked_ticket")
    invoice_items: List["InvoiceItem"] = Relationship(back_populates="linked_ticket")
    quote_items: List["QuoteItem"] = Relationship(back_populates="linked_ticket")

    @property
    def service_fee(self) -> float:
        """Phí dịch vụ – profit margin on this ticket."""
        return self.selling_price - self.net_price

    @property
    def computed_true_income(self) -> float:
        """Thu nhập thực – selling price plus discount minus host net prices."""
        return self.selling_price + self.discount - (
            self.ev_price
            + self.ast_price
            + self.thf_price
            + self.web_price
            + self.insurance_price
        )


# ---------------------------------------------------------------------------
# Request / Response schemas
# ---------------------------------------------------------------------------

class TicketCreate(TicketBase):
    """Payload accepted by POST /tickets (after AI-extraction confirmation)."""

    passengers: List[str] = Field(
        default_factory=list,
        min_length=1,
        description="At least one passenger name is required.",
    )


class TicketRead(TicketBase):
    """Public read representation of a ticket."""

    id: uuid.UUID
    passengers: List[str]
    service_fee: float  # computed, not stored
    created_at: datetime
    updated_at: datetime


class TicketUpdate(SQLModel):
    """All fields optional for partial PATCH payloads."""

    pnr: Optional[str] = Field(default=None, min_length=6, max_length=6)
    airline: Optional[Airline] = None
    ticket_number: Optional[str] = Field(default=None, max_length=50)
    seat_code: Optional[str] = Field(default=None, max_length=20)
    fare_class: Optional[str] = Field(default=None, max_length=50)
    passengers: Optional[List[str]] = None
    departure_place: Optional[str] = Field(default=None, max_length=255)
    arrival_place: Optional[str] = Field(default=None, max_length=255)
    departure_code: Optional[str] = Field(default=None, max_length=10)
    arrival_code: Optional[str] = Field(default=None, max_length=10)
    itinerary: Optional[str] = Field(default=None, max_length=100)
    flight_date: Optional[datetime] = None
    booked_at: Optional[datetime] = None
    net_price: Optional[float] = Field(default=None, ge=0)
    ev_price: Optional[float] = Field(default=None, ge=0)
    ast_price: Optional[float] = Field(default=None, ge=0)
    thf_price: Optional[float] = Field(default=None, ge=0)
    web_price: Optional[float] = Field(default=None, ge=0)
    insurance_price: Optional[float] = Field(default=None, ge=0)
    service_fee: Optional[float] = Field(
        default=None,
        ge=0,
        description="Optional service fee used to recompute selling_price.",
    )
    selling_price: Optional[float] = Field(default=None, ge=0)
    discount: Optional[float] = Field(default=None, ge=0)
    true_income: Optional[float] = None
    status: Optional[TicketStatus] = None
    customer_id: Optional[uuid.UUID] = None

File: web/src/schemas/ticket.ts
/**
 * ticket.ts – Zod schemas for the Ticket domain.
 *
 * Mirrors api/models/ticket.py (TicketCreate, TicketRead, TicketUpdate).
 *
 * Field names from docs/DICTIONARY.md:
 *   pnr           → Mã đặt chỗ (6-char booking reference)
 *   ticket_number → Số vé
 *   departure_place / arrival_place → Nơi đi / nơi đến
 *   departure_code / arrival_code   → Mã nơi đi / mã nơi đến
 *   itinerary     → Hành trình  (e.g. "HAN-SGN")
 *   booked_at     → Ngày xuất vé
 *   net_price     → Giá gốc     (cost paid to airline/supplier)
 *   selling_price → Giá bán     (price invoiced to customer)
 *   discount      → Chiết khấu hãng (airline add-in earned by agency)
 *   ev_price      → Giá net EV (host net price from EV)
 *   ast_price     → Giá AST (host net price from AST)
 *   thf_price     → Giá Thành Hoàng / THF (host net price from Thành Hoàng)
 *   web_price     → Giá WEB (host net price from WEB)
 *   true_income   → Thu nhập thực (selling_price + discount - (ev_price + ast_price + thf_price + web_price))
 *   service_fee   → Phí dịch vụ (computed: selling_price - net_price)
 *
 * Agent parser output (docs/AGENT_PARSER.md) feeds directly into TicketCreateSchema.
 */

import { z } from "zod";
import { calculateServiceFee } from "@/lib/finance-core";
import { AirlineSchema, TicketStatusSchema } from "./enums";

// ---------------------------------------------------------------------------
// Shared base – fields common to all ticket schema variants
// ---------------------------------------------------------------------------

const TicketBaseSchema = z.object({
  /**
   * Mã đặt chỗ – 6-character alphanumeric PNR booking reference.
   * Maps to Python: pnr
   */
  pnr: z
    .string()
    .length(6, "PNR (mã đặt chỗ) must be exactly 6 characters.")
    .toUpperCase(),

  /** Airline carrier code. Maps to Python: airline */
  airline: AirlineSchema.nullable().optional(),

  /** Airline ticket number. Maps to Python: ticket_number */
  ticket_number: z.string().min(1).max(50).nullable().optional(),

  /** Human-readable departure place. Maps to Python: departure_place */
  departure_place: z.string().min(1).max(255).nullable().optional(),

  /** Human-readable arrival place. Maps to Python: arrival_place */
  arrival_place: z.string().min(1).max(255).nullable().optional(),

  /** Departure place code. Maps to Python: departure_code */
  departure_code: z.string().min(1).max(10).toUpperCase().nullable().optional(),

  /** Arrival place code. Maps to Python: arrival_code */
  arrival_code: z.string().min(1).max(10).toUpperCase().nullable().optional(),

  /**
   * Flight route string (hành trình).
   * Maps to Python: itinerary
   * Example: "HAN-SGN" or "SGN-DAD-HAN"
   */
  itinerary: z
    .string()
    .min(1, "Itinerary (hành trình) is required.")
    .max(100),

  /**
   * Scheduled departure datetime in ISO-8601 format.
   * The API stores UTC; UI displays in DD/MM/YYYY per docs/DICTIONARY.md.
   * This is the flight event time and can be in the past.
   * Maps to Python: flight_date
   */
  flight_date: z.coerce.date(),

  /**
   * Real-world datetime when staff booked the ticket.
   * This is separate from flight_date and system created_at.
   * Maps to Python: booked_at
   */
  booked_at: z.coerce.date().nullable().optional(),

  /**
   * Net cost from airline/supplier – giá gốc.
   * Maps to Python: net_price
   */
  net_price: z
    .number({ message: "Net price (giá gốc) is required." })
    .min(0, "Net price (giá gốc) must be ≥ 0."),

  ev_price: z
    .number({ message: "EV net price (giá net EV) is required." })
    .min(0, "EV net price (giá net EV) must be ≥ 0.")
    .default(0),

  ast_price: z
    .number({ message: "AST net price (giá AST) is required." })
    .min(0, "AST net price (giá AST) must be ≥ 0.")
    .default(0),

  thf_price: z
    .number({ message: "THF price (giá Thành Hoàng) is required." })
    .min(0, "THF price (giá Thành Hoàng) must be ≥ 0.")
    .default(0),

  web_price: z
    .number({ message: "WEB price (giá WEB) is required." })
    .min(0, "WEB price (giá WEB) must be ≥ 0.")
    .default(0),

  selling_price: z
    .number({ message: "Selling price (giá bán) is required." })
    .min(0, "Selling price (giá bán) must be ≥ 0."),

  discount: z
    .number({ message: "Airline discount (chiết khấu hãng) is required." })
    .min(0, "Airline discount (chiết khấu hãng) must be ≥ 0.")
    .default(0),

  true_income: z.number({
    message: "True income (thu nhập thực) is required.",
  }),

  /** Lifecycle state of the ticket. Maps to Python: status */
  status: TicketStatusSchema.default("DRAFT"),

  /** Foreign key to the owning Customer. Maps to Python: customer_id */
  customer_id: z.string().uuid("customer_id must be a valid UUID."),
});

// ---------------------------------------------------------------------------
// TicketCreate – POST /tickets
// Aligns 1-to-1 with the JSON returned by the Gemini AI parser.
// ---------------------------------------------------------------------------

export const TicketCreateSchema = TicketBaseSchema.extend({
  /**
   * List of passenger full names in UPPERCASE.
   * At least one passenger is required.
   * Maps to Python: passengers (JSON column)
   */
  passengers: z
    .array(z.string().min(1).toUpperCase())
    .min(1, "At least one passenger is required."),
}).refine(
  (data) =>
    Math.abs(
      data.true_income -
        (data.selling_price +
          data.discount -
          (data.ev_price + data.ast_price + data.thf_price + data.web_price)),
    ) <= 1,
  {
    message:
      "True income (thu nhập thực) must equal selling price + airline discount - EV/AST/THF/WEB net prices.",
    path: ["true_income"],
  }
);

export type TicketCreate = z.infer<typeof TicketCreateSchema>;

// ---------------------------------------------------------------------------
// TicketRead – GET /tickets/:id
// ---------------------------------------------------------------------------

export const TicketReadSchema = TicketBaseSchema.extend({
  /** UUID assigned by the backend. Read-only. */
  id: z.string().uuid(),

  passengers: z.array(z.string()),

  /**
   * Computed on the backend: selling_price - net_price.
   * Phí dịch vụ – not stored in the DB.
   * Maps to Python: service_fee (property)
   */
  service_fee: z.number(),

  /** UTC timestamp of when the ticket record was created. */
  created_at: z.coerce.date(),

  /** UTC timestamp of when the ticket record was last updated. */
  updated_at: z.coerce.date(),
});

export type TicketRead = z.infer<typeof TicketReadSchema>;

// ---------------------------------------------------------------------------
// TicketUpdate – PATCH /tickets/:id (all fields optional)
// ---------------------------------------------------------------------------

export const TicketUpdateSchema = z.object({
  pnr: z.string().length(6).toUpperCase().optional(),
  airline: AirlineSchema.optional(),
  ticket_number: z.string().min(1).max(50).optional(),
  passengers: z.array(z.string().min(1).toUpperCase()).min(1).optional(),
  departure_place: z.string().min(1).max(255).optional(),
  arrival_place: z.string().min(1).max(255).optional(),
  departure_code: z.string().min(1).max(10).toUpperCase().optional(),
  arrival_code: z.string().min(1).max(10).toUpperCase().optional(),
  itinerary: z.string().min(1).max(100).optional(),
  flight_date: z.coerce.date().optional(),
  booked_at: z.coerce.date().nullable().optional(),
  net_price: z.number().min(0).optional(),
  ev_price: z.number().min(0).optional(),
  ast_price: z.number().min(0).optional(),
  thf_price: z.number().min(0).optional(),
  web_price: z.number().min(0).optional(),
  selling_price: z.number().min(0).optional(),
  discount: z.number().min(0).optional(),
  true_income: z.number().optional(),
  status: TicketStatusSchema.optional(),
  customer_id: z.string().uuid().optional(),
});

export type TicketUpdate = z.infer<typeof TicketUpdateSchema>;

// ---------------------------------------------------------------------------
// Derived helpers
// ---------------------------------------------------------------------------

/** Compute phí dịch vụ (service fee) client-side for optimistic UI. */
export const computeServiceFee = (
  net_price: number,
  selling_price: number
): number => calculateServiceFee(net_price, selling_price);

export const computeTrueIncome = (
  selling_price: number,
  discount: number,
  ev_price = 0,
  ast_price = 0,
  thf_price = 0,
  web_price = 0
): number => selling_price + discount - (ev_price + ast_price + thf_price + web_price);
